### PR TITLE
[ASCollectionLayout] Add ASCollectionGalleryLayoutSizeProviding

### PR DESCRIPTION
**Pull request by nguyenhuy**
- This allows users to return different sizes based on certain conditions, such as the collection node's bounds or grid constants.
- ASPagerNode will also act as a size provider to ensure all pages have an up-to-date size that is its bounds.

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: examples/CustomCollectionView/Sample/MosaicCollectionLayoutDelegate.m [hunk 1]
```diff
@@ -36,11 +36,13 @@ - (instancetype)initWithNumberOfColumns:(NSInteger)numberOfColumns headerHeight:
 
 - (ASScrollDirection)scrollableDirections
 {
+  ASDisplayNodeAssertMainThread();
```
Comment (garrettmoon) [inline examples/CustomCollectionView/Sample/MosaicCollectionLayoutDelegate.m]: Are these and the below necessary? Seems a little strange to be in 'client' code? Leftover from testing?

File: Source/Details/ASCollectionGalleryLayoutDelegate.h [hunk 2]
```diff
@@ -23,7 +40,9 @@ NS_ASSUME_NONNULL_BEGIN
 AS_SUBCLASSING_RESTRICTED
 @interface ASCollectionGalleryLayoutDelegate : NSObject <ASCollectionLayoutDelegate>
 
-- (instancetype)initWithScrollableDirections:(ASScrollDirection)scrollableDirections itemSize:(CGSize)itemSize NS_DESIGNATED_INITIALIZER;
```
Comment (garrettmoon) [inline Source/Details/ASCollectionGalleryLayoutDelegate.h]: Does this one need to be deprecated?

Comment (nguyenhuy) [inline Source/Details/ASCollectionGalleryLayoutDelegate.h]: No, it's merged last week and still in Beta.

File: examples/CustomCollectionView/Sample/MosaicCollectionLayoutDelegate.m (hunk 1 above)
Comment (nguyenhuy) [inline examples/CustomCollectionView/Sample/MosaicCollectionLayoutDelegate.m]: I intentionally added to tell whoever reading this example that these methods are called on main. Some rarely read the documentation :) I can definitely remove it if you think it's excessive.

Comment (garrettmoon) [inline examples/CustomCollectionView/Sample/MosaicCollectionLayoutDelegate.m]: Ah good point!

File: Source/Details/ASCollectionGalleryLayoutDelegate.h (hunk 2 above)
Comment (garrettmoon) [inline Source/Details/ASCollectionGalleryLayoutDelegate.h]: 👍

**Comment by nguyenhuy** [comment]
Thanks for the review, @garrettmoon!

**Pull request merged**